Lib/java/std_map.i

```diff
@@ -12,60 +14,188 @@
+      return null;
+    }
+
+    Iterator itr = find(($typemap(jboxtype, KeyType)) key);
+    if (itr.isNot(end())) {
+      return itr.getValue();
+    }
+
+    return null;
+  }
+
+  public $typemap(jboxtype, ValueType) put($typemap(jboxtype, KeyType) key,
+                                           $typemap(jboxtype, ValueType) value) {
+    Iterator itr = find(($typemap(jboxtype, KeyType)) key);
+    if (itr.isNot(end())) {
+      $typemap(jboxtype, ValueType) oldValue = itr.getValue();
+      itr.setValue(value);
+      return oldValue;
+    } else {
+      putUnchecked(key, value);
+      return null;
+    }
+  }
+
+  public $typemap(jboxtype, ValueType) remove(Object key) {
+    if (!(key instanceof $typemap(jboxtype, KeyType))) {
+      return null;
+    }
+
+    Iterator itr = find(($typemap(jboxtype, KeyType)) key);
+    if (itr.isNot(end())) {
+      $typemap(jboxtype, ValueType) oldValue = itr.getValue();
+      removeUnchecked(itr);
+      return oldValue;
+    } else {
+      return null;
+    }
+  }
+
+  public java.util.Set<Entry<$typemap(jboxtype, KeyType), $typemap(jboxtype, ValueType)>> entrySet() {
+    java.util.Set<Entry<$typemap(jboxtype, KeyType), $typemap(jboxtype, ValueType)>> setToReturn =
+        new java.util.HashSet<Entry<$typemap(jboxtype, KeyType), $typemap(jboxtype, ValueType)>>();
+
+    Iterator itr = begin();
+    final Iterator end = end();
+    while(itr.isNot(end)) {
+      setToReturn.add(new Entry<$typemap(jboxtype, KeyType), $typemap(jboxtype, ValueType)>() {
+        private Iterator iterator;
+
+        private Entry<$typemap(jboxtype, KeyType), $typemap(jboxtype, ValueType)> init(Iterator iterator) {
+          this.iterator = iterator;
+          return this;
+        }
+
+        public $typemap(jboxtype, KeyType) getKey() {
+          return iterator.getKey();
+        }
+
+        public $typemap(jboxtype, ValueType) getValue() {
+          return iterator.getValue();
+        }
+
+        public $typemap(jboxtype, ValueType) setValue($typemap(jboxtype, ValueType) newValue) {
+          $typemap(jboxtype, ValueType) oldValue = iterator.getValue();
+          iterator.setValue(newValue);
+          return oldValue;
+        }
+      }.init(itr));
+      itr = itr.getNextUnchecked();
+    }
+
+    return setToReturn;
+  }
+%}
+
+  public:
+    map();
+    map(const map<KeyType, ValueType, Comparator >&);
+
+    struct iterator {
       %extend {
-        const T& get(const K& key) throw (std::out_of_range) {
-          std::map< K, T, C >::iterator i = self->find(key);
-          if (i != self->end())
-            return i->second;
-          else
-            throw std::out_of_range("key not found");
+        std::map<KeyType, ValueType, Comparator >::iterator getNextUnchecked() {
+          std::map<KeyType, ValueType, Comparator >::iterator copy = (*$self);
+          return ++copy;
+        }
+
+        bool isNot(const iterator other) const {
```

Comment:
Pass by reference would be better:
```suggestion
        bool isNot(const iterator &other) const {
```

Reply:
In general C++ iterations are lightweight objects and the STL typically passes them around by value (take a look at the declarations of any container constructor or algorithm that takes in a begin/end iterator range). But up to you, I can change it

Reply:
It's not as simple as that - `const_iterator` and `iterator` are different types - you can't pass an `iterator` into a function that takes `const_iterator`. This struct that this member belongs to is `iterator` not `const_iterator`. To switch to `const_iterator` means that the `setValue` function in the extension part is no longer valid because a `const_iterator` cannot mutate the value it points to.